src/lib/plugin.js

```diff
@@ -160,43 +160,61 @@ class FiveBellsLedger extends EventEmitter2 {
 
-    return new Promise((resolve, reject) => {
-      this.connection = reconnect({immediate: true}, (ws) => {
-        ws.on('open', () => {
-          debug('ws connected to ' + streamUri)
-        })
-        ws.on('message', (msg) => {
-          const notification = JSON.parse(msg)
-          debug('notify transfer', notification.resource.state, notification.resource.id)
-
-          co.wrap(this._handleNotification)
-            .call(this, notification.resource, notification.related_resources)
-            .then(() => {
-              if (this.debugReplyNotifications) {
-                ws.send(JSON.stringify({ result: 'processed' }))
-              }
-            })
-            .catch((err) => {
-              debug('failure while processing notification: ' +
-                (err && err.stack) ? err.stack : err)
-              if (this.debugReplyNotifications) {
-                ws.send(JSON.stringify({
-                  result: 'ignored',
-                  ignoreReason: {
-                    id: err.name,
-                    message: err.message
-                  }
-                }))
-              }
-            })
-        })
-        ws.on('close', () => {
-          debug('ws disconnected from ' + streamUri)
-        })
+    this.connection = reconnect({immediate: true}, (ws) => {
+      ws.on('open', () => {
+        debug('ws connected to ' + streamUri)
+        this.emit('_ws_open')
+      })
+      ws.on('message', (msg) => {
+        const notification = JSON.parse(msg)
+        debug('notify transfer', notification.resource.state, notification.resource.id)
+
+        co.wrap(this._handleNotification)
+          .call(this, notification.resource, notification.related_resources)
+          .then(() => {
+            if (this.debugReplyNotifications) {
+              ws.send(JSON.stringify({ result: 'processed' }))
+            }
+          })
+          .catch((err) => {
+            debug('failure while processing notification: ' +
+              (err && err.stack) ? err.stack : err)
+            if (this.debugReplyNotifications) {
+              ws.send(JSON.stringify({
+                result: 'ignored',
+                ignoreReason: {
+                  id: err.name,
+                  message: err.message
+                }
+              }))
+            }
+          })
+      })
+      ws.on('error', (e) => {
+        this.emit('_ws_error', new Error('websocket connection error'))
+      })
+      ws.on('close', () => {
+        debug('ws disconnected from ' + streamUri)
+        this.emit('_ws_error', new Error('websocket disconnected'))
+      })
 
-        // reconnect-core expects the disconnect method to be called: `end`
-        ws.end = ws.close
+      // reconnect-core expects the disconnect method to be called: `end`
+      ws.end = ws.close
+    })
+
+    // events are required here due to the nested structure of
+    // this.connection and the websocket exposed only through
+    // its callback
```

Comment:
I need to listen for the events from outside of the promise because I need to return `wsOpen` in `Promise.all`. If I wrapped the `reconnect` call inside of a promise, then there would be no guarantee that `this.connection` is set when it registers the event listeners on connection right afterwards